sample_consensus/include/pcl/sample_consensus/impl/sac_model_circle.hpp

```diff
@@ -64,19 +64,23 @@ pcl::SampleConsensusModelCircle2D<PointT>::isSampleGood(const Indices &samples)
   // Compute the segment values (in 2d) between p2 and p0
   p2 -= p0;
 
-  Eigen::Array2d dy1dy2 = p1 / p2;
+  // Check if the triangle area spanned by the three sample points is greater than 0
+  // https://en.wikipedia.org/wiki/Triangle#Using_vectors
+  if (std::abs (p1.x()*p2.y() - p2.x()*p1.y()) <= std::numeric_limits<double>::epsilon ()) {
```

Comment:
Yeah, I guess 2e-16 is probably too low. Eigen itself uses a self-defined `dummy_precision` ([1e-5 for float](https://gitlab.com/libeigen/eigen/-/blob/d49ede4dc4203942c2e01a7b39f23f2d9710765e/Eigen/src/Core/NumTraits.h#L237), [1e-12 for double](https://gitlab.com/libeigen/eigen/-/blob/d49ede4dc4203942c2e01a7b39f23f2d9710765e/Eigen/src/Core/NumTraits.h#L243)) as default, e.g. in [`isZero()`](https://gitlab.com/libeigen/eigen/-/blob/d49ede4dc4203942c2e01a7b39f23f2d9710765e/Eigen/src/Core/DenseBase.h#L376). So maybe that's something to make use of?

Otherwise something between 1e-5 and 1e-7 could be the way to go. 1e-5/1e-6 would leave about two/one order(s) of magnitude of margin to compute something with floats and ample margin for double :laughing:

Reply:
Right, `dummy_precision` is also an option. Let's use that.
BTW let's try to not make this PR much bigger. The bigger, the more difficult it is to review and the longer it takes. If there are many more changes you want to make, maybe consider creating another PR for that?

Reply:
Sorry to bother you once again, but would you like to see`Eigen::NumTraits<double>::dummy_precision ()` or `Eigen::NumTraits<float>::dummy_precision ()` here (and in Circle3D)?

With Circle and Circle3D, this should be fairly "complete" from what I can contribute. The only thing I could offer is to change the checks in the other models from `std::numeric_limits<float>::epsilon ()` to `Eigen::NumTraits<double>::dummy_precision ()` as well, but that's about it.

Other than that, please let me know if you want the commits to be squashed (e.g. per model) or if you want to do squash-merge all of them. (And before I forget: thanks for your patience sticking with me on this adventure so far.)

Reply:
> Sorry to bother you once again, but would you like to see`Eigen::NumTraits<double>::dummy_precision ()` or `Eigen::NumTraits<float>::dummy_precision ()` here (and in Circle3D)?

Hm, difficult to decide. Maybe double, if all computation in `isSampleGood` and `computeModelCoefficients` is in double (1e-12 seems reasonable then), else float?

> With Circle and Circle3D, this should be fairly "complete" from what I can contribute. The only thing I could offer is to change the checks in the other models from `std::numeric_limits<float>::epsilon ()` to `Eigen::NumTraits<double>::dummy_precision ()` as well, but that's about it.

I would say changing from epsilon to dummy_precision everywhere is not necessary if we are not changing those code parts anyway.

> Other than that, please let me know if you want the commits to be squashed (e.g. per model) or if you want to do squash-merge all of them. (And before I forget: thanks for your patience sticking with me on this adventure so far.)

One commit with the new tests, one commit with the changes in sample_consensus/* would be nice. Thank you for your contribution!